### ISSUE TITLE
Category-level Consistency (CC) computation

### ISSUE THREAD

**Issue by Kaisan-Li** [brown-ivl/ConDor#3]
Hi. Thank you for sharing codes and resources.
I'm a bit confused about how you compute the CC metric. Following your description in the paper, I write a pseudo-code below:
```python
# Assume we have a canonicalized P_cano_set with shape [M N 3], where M is the number of shapes and N is the number of points of each shape. 
# Here P_cano_set is a set of <num>_canonical_pointcloud_full_splits.ply which you obtain after running the test script.

total_loss = 0
for i in range(M):
    choice = np.random.choice(M, 120, replace=False)
    P_i = P_cano_set[i].unsqueeze(0).repeat(120,1,1) #[120 N 3]
    P_set = P_cano_set[choice] #[120 N 3]
    dist = chamfer_dist(P_i, P_set) #[120]
    total_loss += mean(dist)
    
avg_loss = total_loss/M
```  

But the metrics computed this way are different from the ones you reported in the paper. For example, the CC metric of the bench (Full) is 0.0343, differing from 0.1043 in the paper.

Could you point out the mistakes in my understanding? Or could you introduce a bit of how to run the codes in /evaluators folder? Thank you.

**Comment by RahulSajnani** [comment]
Hi @Kaisan-Li 

Thank you for showing interest in out work! 
The category-level canonicalization is defined correctly by you. The few differences in our case are as follows:
1. We pre-compute the random choice vector and keep it fixed while evaluation so that the metrics do not change on every run.
2. We also randomly rotate each shape by P number of rotations

In the class consistency metric [code](https://github.com/brown-ivl/ConDor/blob/f387b9211b0f663f3d8c6d3ad4e58010ffbac3e2/ConDor/evaluators/canonicalization_metrics.py#L530-L561), we pass the following arguments:
1. filename - filename for h5 pointclouds file
2. x_path - path to the parent directory that have pointclouds
3. r_input_path - path to random rotations h5 file
4. r_can_path - path to h5 file with the predicted canonicalizing rotation matrices
5. idx_shapes_path - precomputed random indexes for every shape in the dataset 
6. idx_rots_path - precomputed random indexes for applying rotations

In your script I will make the following changes:

```python

# Assume we have a canonicalized P_cano_set with shape [M P N 3], where M is the number of shapes, N is the number of points of each shape, and P is the number of random rotations applied before canonicalization for each shape. 
# Here P_cano_set is a set of <num>_canonical_pointcloud_full_splits.ply which you obtain after running the test script.

total_loss = 0
for i in range(M):
    choice = np.random.choice(M, 120, replace=False) # fixed before running evaluation
    choice_rotation = np.random.choice(P, 64, replace=False) # fixed before evaluation
    for j in range(P)
        P_i = P_cano_set[i][choice_rotation].unsqueeze(0).repeat(120,1, 1,1) #[120 P N 3]
        P_set = P_cano_set[choice][choice_rotation] #[120 P N 3]
    dist = chamfer_dist(P_i, P_set) #[120, P]
    total_loss += mean(dist)
    
avg_loss = total_loss/M
```

**Comment by Kaisan-Li** [comment]
Hi @RahulSajnani 
Thank you for the illustration! It's much clear now. So I think your point is that a shape will be canonicalized to different poses according to its rotation. Regarding the number of rotations, do you set P=120 following the setting in the IC metric?

I also want to make sure about the details of IC and GC metrics, could you see if there are any mistakes in the following codes:
```python
# IC computation
# Assume we have a pre-aligned P_set with shape [M N 3], where M is the number of shapes, N is the number of points of each shape
# And we have a pre-defined R_set with shape [120 3 3]
# And we have a model() which takes as input a point cloud [B N 3] and outputs its canonicalized version [B N 3]

total_loss = 0
for i in range(M):
    P_i = P_set[i] #[N 3]
    P_i_cano = model(P_i) #[N 3]
    P_i_rot = torch.matmul(P_i.unsqueeze(0).repeat(120,1,1), R_set) #[120 N 3]
    P_i_rot_cano = model(P_i_rot) #[120 N 3]
    
    cd_i = chamfer_dist(P_i_cano.unsqueeze(0).repeat(120,1,1), P_i_rot_cano) #[120]
    total_loss += mean(cd_i)
    
avg_loss = total_loss/M
```  
  
```python
# GC computation
# Assume we have a pre-aligned P_set with shape [M N 3], where M is the number of shapes, N is the number of points of each shape
# And we have a model() which takes as input a point cloud [B N 3] and outputs its canonicalizing frame [B 3 3]

total_loss = 0
for i in range(M):
    for j in range(M):
        for k in range(M):
            x_i = P_set[i] 
            x_j = P_set[j]
            x_k = P_set[k]
            x_j_frame = model(x_j)
            x_k_frame = model(x_k)
            dist = chamfer_dist(torch.matmul(x_i, x_j_frame), torch.matmul(x_i, x_k_frame)) #[1]
            total_loss += dist
            
avg_loss = total_loss/M**3
```
  
Thank you so much for your time.

**Comment by RahulSajnani** [comment]
Hi @Kaisan-Li
I made a small error in the pseudo-code. The corrected one is given below:

```python3
total_loss = 0
for i in range(M):
    choice_shape = np.random.choice(M, 120, replace=False) # fixed before running evaluation
    choice_rotation = np.random.choice(P, 64, replace=False) # fixed before evaluation
    P_i = P_cano_set[i][choice_rotation].unsqueeze(0).repeat(120,1, 1,1) #[120 P N 3]
    P_set = P_cano_set[choice_shape][choice_rotation] #[120 P N 3]
    dist = chamfer_dist(P_i, P_set) #[120, P]
    total_loss += mean(dist)
    
avg_loss = total_loss/M

```
I will get back to you on the remaining metrics by tomorrow. Sorry for the delay

**Comment by RahulSajnani** [comment]
Hi @Kaisan-Li 
You can check [this link](https://drive.google.com/drive/folders/150kKsqvH26tqg3yReh9e2HUDMyCuMAqZ?usp=sharing) for rotations, rotation idx, shape permutations, and the dataset to evaluate your code.

```python3
# IC computation
# IC is for checking canonicalization for multiple rotations but for the same object instance
# And we have a pre-defined R_set_1 with shape [120 3 3] and R_set_2 with shape [120 3 3]
# And we have a model() which takes as input a point cloud [B N 3] and outputs its canonicalized version [B N 3]

total_loss = 0
for i in range(M):
    P_i = P_set[i] #[N 3]
    # P_i_cano = model(P_i) #[N 3] commented this line as we don't assume a canonicalized dataset for this metric
    P_i_rot_1 = torch.matmul(P_i.unsqueeze(0).repeat(120,1,1), R_set_1)
    P_i_rot_2 = torch.matmul(P_i.unsqueeze(0).repeat(120,1,1), R_set_2) #[120 N 3]
    P_i_rot_cano_1 = model(P_i_rot_1) #[120 N 3]
    P_i_rot_cano_2 = model(P_i_rot_2) #[120 N 3]
    
    cd_i = chamfer_dist(P_i_rot_cano_1, P_i_rot_cano_2) #[120]
    total_loss += mean(cd_i)
    
avg_loss = total_loss/M
```

For the GC metric, please check the code below. Even though your idea is correct, we implement this metric in a slightly different way in the code.
```python3
# GC computation
# Assume we have a pre-aligned P_set with shape [M N 3], where M is the number of shapes, N is the number of points of each shape
# And we have a model() which takes as input a point cloud [B N 3] and outputs its canonicalizing frame [B 3 3]
# We have shape_idx of dimenstion [M P]. P = 64 typically in our experiments. Higher P value gives more robust metrics


total_loss = 0
for i in range(M):
   x_i = P_set[i] # N, 3
   x_jk = P_set[shape_idx[i]] # [P N 3] these are the other shapes we compare with
   x_jk_frame = model(x_jk) # [P 3 3]
   x_i_frame = model(x_i) # [1 3 3]
   x_jk_cano = matmul(x_jk_frame, x_i) # [P N 3]
   x_i_cano = matmuil(x_i_frame, x_i) # [P N 3]
   dist = chamfer_dist(x_i_cano, x_jk_cano) #[P]
   total_loss += mean(dist)
    
avg_loss = total_loss/M
```

I hope this clears your queries on evaluating the IC and GC metrics
Note that IC and CC metrics do not assume pre-canonicalized data. The GC metric assumes canonicalized data. Please let me know if you have any other queries. I have tried to follow your coding style, but let me know if something is not clear.

**Comment by Kaisan-Li** [comment]
Hi @RahulSajnani 
Thank you for providing the data, it's quite helpful.
I have some questions about your [CC metric](https://github.com/brown-ivl/ConDor/blob/main/ConDor/evaluators/canonicalization_metrics.py#L482-L521):  
1. I can tell that 'r_input' is a [128 3 3] tensor loaded from rotations.h5.  And I think r_can is a 4-D tensor like [M 128 3 3] since you used tf gather() twice to get shapes and rotations respectively. So will line 483 output 3? It looks strange...
2. In the following loop, it seems that you 
 (1) select a rotation in r_input
 (2) select two batches from ordered shapes and permutated shapes respectively, then apply rotation and canonicalization
 (3) compute chamfer distance
 I think it's different from the CC definition.

Could you point out my mistakes?

**Comment by RahulSajnani** [comment]
Hi @Kaisan-Li 
There was a small typo at my end. In GC metric, all the matrices are multiplied with x_i only and compared with each other. I edited my last comment to fix it

**Comment by RahulSajnani** [comment]
> Hi @RahulSajnani Thank you for providing the data, it's quite helpful. I have some questions about your [CC metric](https://github.com/brown-ivl/ConDor/blob/main/ConDor/evaluators/canonicalization_metrics.py#L482-L521):
> 
>     1. I can tell that 'r_input' is a [128 3 3] tensor loaded from rotations.h5.  And I think r_can is a 4-D tensor like [M 128 3 3] since you used tf gather() twice to get shapes and rotations respectively. So will line 483 output 3? It looks strange...
> 
>     2. In the following loop, it seems that you
>        (1) select a rotation in r_input
>        (2) select two batches from ordered shapes and permutated shapes respectively, then apply rotation and canonicalization
>        (3) compute chamfer distance
>        I think it's different from the CC definition.
> 
> 
> Could you point out my mistakes?

The code similar to the definition in the paper. We take two randomly oriented batch of shapes and canonicalize them. After canonicalization, we compute the Chamfer Distance between them. This is the category-level consistency equation in the paper:
$$CC := mean(CD(X_i^c, X_j^c))$$ where i and j are shape ids for two different shapes and c refers to the shapes being canonicalized.
We apply random transforms in the code because the dataset is pre-canonicalized. But the CC metric does not assume canonicalization.

**Comment by Kaisan-Li** [comment]
Hi @RahulSajnani 
Thank you! I'll have a look at IC and GC later.
On a separate note, I intend to get some canonicalized shapes and the corresponding frames from your pre-trained model. I think in the 2 lines of [codes](https://github.com/brown-ivl/ConDor/blob/main/ConDor/trainers/trainer_main.py#L304-L307), output_dict["x"] is the randomly rotated input, and "basis_instance_to_canonical" is the canonicalizing frame. I can use either 
 `tf.einsum('bij,bkj->bki', basis_instance_to_canonical, output_dict["x"])`  
or  
 `tf.matmul(output_dict["x"], basis_instance_to_canonical)`    
to get canonicalized shapes. Is that correct?

**Comment by RahulSajnani** [comment]
Hi @Kaisan-Li 
No it should be 
`tf.einsum('bij,bkj->bki', basis_instance_to_canonical, output_dict["x"])` 
or 
`tf.matmul(output_dict["x"], tf.transpose(basis_instance_to_canonical, perm=[0, 2, 1]))`. Please check with a small python example at your end

**Comment by Kaisan-Li** [comment]
Hi @RahulSajnani 
Thank you! I have no further questions. Thanks again for your kind help.
